modules/pagerduty/pagerduty.go

```diff
@@ -0,0 +1,188 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Contributor: Miles Crabill <mcrabill@mozilla.com>
+
+package pagerduty
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	pagerduty "github.com/PagerDuty/go-pagerduty"
+
+	"go.mozilla.org/userplex/modules"
+)
+
+func init() {
+	modules.Register("pagerduty", new(module))
+}
+
+type module struct{}
+
+func (m *module) NewRun(c modules.Configuration) modules.Runner {
+	r := new(run)
+	r.Conf = c
+	return r
+}
+
+type run struct {
+	Conf modules.Configuration
+	p    parameters
+	c    credentials
+
+	pagerdutyEmailToLdapEmail  map[string]string
+	ldapEmailToPagerdutyEmails map[string][]string
+}
+
+type parameters struct {
+	Subdomain string
+}
+
+type credentials struct {
+	APIKey string
+}
+
+func (r *run) Run() (err error) {
+	err = r.Conf.GetParameters(&r.p)
+	if err != nil {
+		return
+	}
+	err = r.Conf.GetCredentials(&r.c)
+	if err != nil {
+		return
+	}
+
+	r.buildLdapMapping()
+	ldapers := r.getLdaperEmails()
+
+	client := pagerduty.NewClient(r.c.APIKey)
+
+	var allUsers []pagerduty.User
+	offset := uint(0)
+	opt := pagerduty.ListUserOptions{
+		APIListObject: pagerduty.APIListObject{Offset: offset},
+	}
+	for {
+		usersResp, err := client.ListUsers(opt)
+		if err != nil {
+			log.Fatalf("[error] pagerduty %s: %s", r.p.Subdomain, err.Error())
+		}
+		allUsers = append(allUsers, usersResp.Users...)
+		if !usersResp.More {
+			break
+		}
+		offset += usersResp.Limit
+		opt.APIListObject.Offset = offset
+	}
+
+	log.Printf("[info] pagerduty %s: found %d users", r.p.Subdomain, len(allUsers))
+
+	// all pagerduty users
+	for _, user := range allUsers {
+		// if there's an ldapuid/localuid mapping for the user
+		var ldapEmail, ldapEmailString string
+		if email, ok := r.pagerdutyEmailToLdapEmail[user.Email]; ok {
+			ldapEmail = email
+			ldapEmailString = " (pagerduty) / " + ldapEmail + " (ldap)"
+		}
+		if ldapEmail == "" {
+			ldapEmail = user.Email
+		}
+
+		// if the user is in ldap
+		if _, ok := ldapers[ldapEmail]; ok {
+			ldapers[ldapEmail] = true
+		} else {
+			// user not in ldap
+			log.Printf("[info] pagerduty %s: user %s%s is not in ldap groups %s but is in PagerDuty account %s", r.p.Subdomain, user.Email, ldapEmailString, r.Conf.LdapGroups, r.p.Subdomain)
+			if r.Conf.Delete {
+				if !r.Conf.ApplyChanges {
+					log.Printf("[dryrun] pagerduty %s: would have deleted user %s", r.p.Subdomain, user.Email)
+				} else {
+					err := client.DeleteUser(user.ID)
+					if err != nil {
+						log.Fatalf("[error] pagerduty %s: could not delete user %s: %s", r.p.Subdomain, user.Email, err.Error())
+					}
+				}
+				r.notify(ldapEmail, fmt.Sprintf("Userplex deleted %s from Pagerduty account %s", ldapEmail, r.p.Subdomain))
```

Comment:
I like to use `%q` instead of `%s` because it automatically quotes strings, which is useful to detect empty variables.